lib/compiler-singlepass/src/compiler.rs

```diff
@@ -106,29 +106,26 @@ impl Compiler for SinglepassCompiler {
             .map(|(i, input)| {
                 let reader = wasmer_compiler::FunctionReader::new(input.module_offset, input.data);
 
-                let mut local_reader = reader.get_locals_reader()?;
-                // This local list excludes arguments.
-                let mut locals = vec![];
-                let num_locals = local_reader.get_count();
-                for _ in 0..num_locals {
-                    let (count, ty) = local_reader.read()?;
-                    for _ in 0..count {
-                        locals.push(ty);
-                    }
-                }
-
                 let mut generator = FuncGen::new(
                     module,
                     module_translation,
                     &self.config,
                     &vmoffsets,
                     &table_styles,
                     i,
-                    &locals,
                     calling_convention,
                 )
                 .map_err(to_compile_error)?;
 
+                let mut local_reader = reader.get_locals_reader()?;
+                for _ in 0..local_reader.get_count() {
+                    let (count, ty) = local_reader.read()?;
+                    // Overflowing has been already been validated by the validator.
+                    generator.feed_local(count, ty).map_err(to_compile_error)?;
+                }
+
+                generator.emit_head().map_err(to_compile_error)?;
```

Comment:
So, the reason `emit_head` moved to here was to enable us to call `emit_head` _after_ the locals have been “fed” which is required for `emit_head` to reserve a correct amount of stack space. We could've kept both operations inside the `FuncGen::new`, but that would then require doing something very smart to the signature of `FuncGen::new`, so that the following are true:

1. We can parse the wasm incrementally still; and
2. We can return wasm parse errors as `CompileError` (and not `CodegenError`).

Reply:
Oh ok now I think I understand why this happens this way!